### PR TITLE
Misc fixes/tuning for transpose scheduler

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- Add `TransposeParams` support for our benchmarking utilities
- The transpose scheduler can now be enabled by env `PYTORCH_NVFUSER_ENABLE` instead of the compile-time flag. This will be much more convenient for debugging.
- View is rejected in `canScheduleCompileTime`
- Some fixes for vectorization factor calculation

The transpose scheduler is still not enabled by default because I am still seeing `CUDA driver error: misaligned address` in `test_norm_large`. Strangely, the stats and inputs are:
```
===== Transpose Stats ========
inputs: T0_g[ iS0{i1}, iS1{i2}, iS2{i3} ], T18_g[ bS30{1}, iS31{i2}, bS32{1} ], T3_g[ rS5{i1}, iS6{i2}, rS7{i3} ]
outputs: T20_g[ iS36{i1}, iS37{i2}, iS38{i3} ]
num_elems: 268435456
n_input_tensors: 3
max_input_dtype_size: 4
group 1: T20_g[ iS36{i1}, iS37{i2}, iS38{i3} ], T0_g[ iS0{i1}, iS1{i2}, iS2{i3} ]
group 2: T18_g[ bS30{1}, iS31{i2}, bS32{1} ], T3_g[ rS5{i1}, iS6{i2}, rS7{i3} ]


===== Transpose Parameters ========
Tag: Transpose heuristics Transpose Characteristics:
 Gridx: 1 BlckY: 1 BlckX: 128
 input tile size: 32
 output tile size: 32
 elements per tile: 1024
 elements per thread: 8
Vectorize set 1, Factor: 4
Unroll set 1, Factor: 2
Vectorize set 2, Factor: 4
Unroll set 2, Factor: 2
====================================

Arguments for kernel10:
Inputs:
  Float [512, 1024, 512] (strides = [524288, 512, 1], address = 0x7f8916000000)
  Float [1, 1024, 1] (strides = [1024, 1, 1], address = 0x7f8adce08200)
  Float [1024] (strides = [1], address = 0x7f8adce09200)
Outputs:
  Float [512, 1024, 512] (strides = [524288, 512, 1], address = 0x7f88d6000000)

```
This sounds totally vectorizable to me, I don't know why there is a misaligned address access. Looks like the following index is problematic:
```CUDA
&T3[((((((nvfuser_index_t)blockIdx.x) % ((ceilDiv(T0.size[1], 32)) * (ceilDiv(T0.size[2], 32)))) / (ceilDiv(T0.size[2], 32))) * 32) + (((((i189 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 4) / 32))]
```
There is a `/32` which could lead to addresses not aligned to 4. I am not sure if this is another indexing bug, I will definitely look into it, but for this PR, I will just leave it unfixed (transpose scheduler is still not enabled by default, so we should be fine).

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 1]
```diff
@@ -126,13 +126,15 @@ class DomainMap : public pointwise_utils::DomainMap {
...
-              grouped.count(member_tv) == 0 || member_tv == tv,
-              "The group of ",
-              member_tv->toString(),
-              " is ambiguous. This is likely a bug.");
-          grouped.emplace(member_tv);
-          groups.back().emplace_back(member_tv);
+          if (grouped.count(member_tv) == 0) {
+            grouped.emplace(member_tv);
+            groups.back().emplace_back(member_tv);
+          } else if (member_tv != tv) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:132]: Are there cases where you expect this to happen or you're just deciding not to error in case it does but just return not supported?

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 2]
```diff
@@ -298,41 +301,19 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
-      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size1);
-  expanded_vector_word_size2 = std::min(
-      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size2);
-
-  vectorize_factor1 = std::max(vectorize_factor1, expanded_vector_word_size1);
-  vectorize_factor2 = std::max(vectorize_factor2, expanded_vector_word_size2);
-
-  params->vectorize_factor1 = vectorize_factor1;
-  params->vectorize_factor2 = vectorize_factor2;
+  params->vectorize_factor1 =
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp]: Why did you remove `expandVectorizationToContigMergedDomains`? Is this because you're not just doing contiguous merges on both of the groups? I guess this could be re-added for the merged domains for small dim cases (though may not be too important for perf right now).

**Review by csarofeen — COMMENTED** [review]
This looks good but going to block on https://github.com/csarofeen/pytorch/pull/1910#discussion_r951463237

Also wondering if you should also enable: https://github.com/csarofeen/pytorch/pull/1882 on this env flag.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:132]: The transpose scheduler is currently on the top of `hlist`, so this check will run on all fusions. And this does happen for some fusion that should be scheduled by the persistent scheduler. I don't know why though.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp]: I don't think this can be re-added. For example,
```
T0[I0, I1, I2, I3] reference
T1[I0, I2, I1, I3]
```
For this case, `T0` and `T1` will be in the same group, but the vectorization factor can not be expanded even with the "virtual innermost dim".

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp]: Gotcha, yeah it would require having a virtual inner dim that's consistent in the group.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:132]: Is this from aliasing inputs to outputs on BN kernels?

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:132]: There are many of them:
```
[  PASSED  ] 685 tests.
[  FAILED  ] 14 tests, listed below:
[  FAILED  ] NVFuserTest.FusionMagicSchedulerLayerNormBackward_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerRMSNormBackward_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerBatchNormalization_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerInstanceNormalization_CUDA
[  FAILED  ] NVFuserTest.FusionSegmentReduceSoftmax_CUDA
[  FAILED  ] NVFuserTest.FusionBNBackwardRepro_CUDA
[  FAILED  ] NVFuserTest.FusionBNRepro_CUDA
[  FAILED  ] NVFuserTest.FusionBNRepro2_CUDA
[  FAILED  ] NVFuserTest.FusionTranslate2Welford_CUDA
[  FAILED  ] NVFuserTest.FusionForceFp16NotAllCast_CUDA
[  FAILED  ] NVFuserTest.FusionForceBf16NotAllCast_CUDA
[  FAILED  ] NVFuserTest.FusionViewPersistentShmoo_CUDA
[  FAILED  ] NVFuserTest.FusionViewConcreteDomain2_CUDA
```

**Comment by zasdfgbnm** [comment]
After merging in https://github.com/csarofeen/pytorch/pull/1910, I see a failure in `NVFuserTest.FusionViewReductionShmoo_CUDA` when running with `PYTORCH_NVFUSER_ENABLE="transpose_scheduler"`. The error seems to be the same as https://github.com/csarofeen/pytorch/issues/1873. I will merge this PR as is, but the issue needs to be fixed before enabling the transpose scheduler by default.
